tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -850,5 +850,523 @@ int main(int argc, char **argv)
+                };
+                count = sizeof(wire_aes_256_and_chacha20) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                /* Verify that client negotiates server preferred non-chacha20 aes_256 when client did not signal */
+                /* cppcheck-suppress redundantAssignment */
+                cipher_preferences.allow_chacha20_boosting = true;
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, wire_aes_256_and_chacha20, count));
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_tls13_aes_256_gcm_sha384);
+
+                EXPECT_SUCCESS(s2n_disable_tls13_in_test());
+            }
+
+            /* Server can negotiate chacha20 ciphersuite, even when boosting is disabled */
+            {
+                DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(connection);
+                EXPECT_SUCCESS(s2n_enable_tls13_in_test());
+
+                connection->security_policy_override = &security_policy;
+                connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+                EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(connection, S2N_TLS13));
+
+                static struct s2n_cipher_suite *test_cipher_suite_list[] = {
+                    /* Not negotiated (not in client wire) */
+                    &s2n_tls13_aes_128_gcm_sha256,                  
+                    /* Not negotiated (not in client wire) */
+                    &s2n_tls13_aes_256_gcm_sha384,                  
+                    /* Negotiated (only available option) */
+                    &s2n_tls13_chacha20_poly1305_sha256,            
+                };
+
+                cipher_preferences = (struct s2n_cipher_preferences) {
+                    .count = s2n_array_len(test_cipher_suite_list),
+                    .suites = test_cipher_suite_list,
+                    .allow_chacha20_boosting = true,
+                };
+
+                /*
+                 * Client signalled chacha20 boosting.
+                 * TLS_CHACHA20_POLY1305_SHA256 is the only option
+                 */
+                uint8_t chacha20_only_wire[] = {
+                    TLS_CHACHA20_POLY1305_SHA256,
+                };
+                uint8_t count = sizeof(chacha20_only_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                /* Verify server can still negotiate chacha20 if it's the only option even when chacha20 boosting is off */
+                cipher_preferences.allow_chacha20_boosting = false;
+                EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(connection, chacha20_only_wire, count));
+                EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_tls13_chacha20_poly1305_sha256);
+                EXPECT_SUCCESS(s2n_disable_tls13_in_test());
+            }
+
+            /* sslv2 server correctly negotiates ciphersuite when chacha20 boosting is enabled */
```

Comment:
Probably worth explaining this case in more detail because it's very non-obvious behavior. Like:
```suggestion

            /* sslv2 server correctly negotiates ciphersuite when chacha20 boosting is enabled
             *
             * <Explain why we're chosing an AES cipher even though chacah20 boosting is enabled and triggered>
             */
```